### ISSUE TITLE
No need to have two gradient buffers for mixed precision

### ISSUE THREAD

**Issue by siddharth9820** [hpcgroup/axonn#23]
Have an argument ``--grad-acc-dtype`` which can be ``fp16`` or ``fp32``.

## ``--grad-acc-dtype fp16``
- Do not create a fp-32 grad buffer
- Split optimizer into param groups
- Do upscaling and optimizer step for each param group separately

Essentially, this would imply multiple calls to optimizer.step()

## ``--grad-acc-dtype fp32``
- Attach backward hooks to params
- The hook should accumulate gradient into a preallocated fp-32 grad buffer and delete the fp-16 grad
